cpp/src/rolling/rolling.cu

```diff
@@ -0,0 +1,369 @@
+  {
+    T val = agg_op::template identity<T>();
+    // declare this as volatile to avoid some compiler optimizations that lead to incorrect results
+    // for CUDA 10.0 and below (fixed in CUDA 10.1)
+    volatile cudf::size_type count = 0;
+
+    size_type window = window_begin[i];
+    size_type forward_window = forward_window_begin[i];
+
+    // compute bounds
+    size_type start_index = max(0, i - window + 1);
+    size_type end_index = min(input.size(), i + forward_window + 1); // exclusive
+
+    // aggregate
+    // TODO: We should explore using shared memory to avoid redundant loads.
+    //       This might require separating the kernel into a special version
+    //       for dynamic and static sizes.
+    for (size_type j = start_index; j < end_index; j++) {
+      if (!has_nulls || input.is_valid(j)) {
+        // Element type and output type are different for COUNT
+        T element = (op == rolling_operator::COUNT) ? T{0} : input.element<T>(j);
+        val = agg_operator(element, val);
+        count++;
+      }
+    }
+
+    // check if we have enough input samples
+    bool output_is_valid = (count >= min_periods);
+
+    // set the mask
+    cudf::bitmask_type result_mask{__ballot_sync(active_threads, output_is_valid)};
+
+    // only one thread writes the mask
+    if (0 == threadIdx.x % cudf::experimental::detail::warp_size) {
+      output.set_mask_word(cudf::word_index(i), result_mask);
+      warp_valid_count += __popc(result_mask);
+    }
+
+    // store the output value, one per thread
+    if (output_is_valid)
+      cudf::detail::store_output_functor<T, op == rolling_operator::MEAN>{}(output.element<T>(i),
+                                                                            val, count);
+
+    // process next element 
+    i += stride;
+    active_threads = __ballot_sync(active_threads, i < input.size());
+  }
+
+  // sum the valid counts across the whole block  
+  gdf_size_type block_valid_count = 
+    cudf::experimental::detail::single_lane_block_sum_reduce<block_size, 0>(warp_valid_count);
+  
+  if(threadIdx.x == 0) {
+    atomicAdd(output_valid_count, block_valid_count);
+  }
+}
+
+struct rolling_window_launcher
+{
+  template<typename T, typename agg_op, rolling_operator op, typename WindowIterator,
+    std::enable_if_t<cudf::detail::is_supported<T, agg_op, 
+                                                op == rolling_operator::MEAN>()>* = nullptr>
+  std::unique_ptr<column> dispatch_aggregation_type(column_view const& input,
+                                                    WindowIterator window_begin,
+                                                    WindowIterator forward_window_begin,
+                                                    size_type min_periods,
+                                                    rmm::mr::device_memory_resource *mr,
+                                                    cudaStream_t stream)
+  {
+    if (input.size() == 0) return empty_like(input);
```

Comment:
Should `column_view` have an `is_empty` method? Obviously wouldn't be done in this PR

Reply:
Adding in this PR.